src/tests/fixing.rs

```diff
@@ -36,124 +31,32 @@ fn test_single_fix() {
     );
 }
 
-struct IdentifierReplacingRule {
-    name: String,
-    replacement: String,
-    listener_queries: Vec<RuleListenerQuery>,
-}
-
-impl IdentifierReplacingRule {
-    pub fn new(name: String, replacement: String) -> Self {
-        Self {
-            listener_queries: vec![RuleListenerQuery {
-                query: format!(
-                    r#"(
-                      (identifier) @c (#eq? @c "{}")
-                    )"#,
-                    name
-                ),
-                capture_name: None,
-            }],
-            name,
-            replacement,
-        }
-    }
-}
-
-impl Rule for IdentifierReplacingRule {
-    fn meta(&self) -> RuleMeta {
-        RuleMeta {
-            name: format!("replace_{}_with_{}", self.name, self.replacement),
-            fixable: true,
-            languages: vec![SupportedLanguage::Rust],
-        }
-    }
-
-    fn listener_queries(&self) -> &[RuleListenerQuery] {
-        &self.listener_queries
-    }
-
-    fn instantiate(&self, _config: &Config) -> Arc<dyn RuleInstance> {
-        Arc::new(IdentifierReplacingRuleInstance::new(
-            self.name.clone(),
-            self.replacement.clone(),
-        ))
-    }
-}
-
-struct IdentifierReplacingRuleInstance {
-    name: String,
-    replacement: String,
-}
-
-impl IdentifierReplacingRuleInstance {
-    fn new(name: String, replacement: String) -> Self {
-        Self { name, replacement }
-    }
-}
-
-impl RuleInstance for IdentifierReplacingRuleInstance {
-    fn instantiate_per_file(&self, _file_run_info: &FileRunInfo) -> Arc<dyn RuleInstancePerFile> {
-        Arc::new(IdentifierReplacingRuleInstancePerFile::new(
-            self.name.clone(),
-            self.replacement.clone(),
-        ))
-    }
-}
-
-struct IdentifierReplacingRuleInstancePerFile {
-    name: String,
-    replacement: String,
-}
-
-impl IdentifierReplacingRuleInstancePerFile {
-    fn new(name: String, replacement: String) -> Self {
-        Self { name, replacement }
-    }
-}
-
-impl RuleInstancePerFile for IdentifierReplacingRuleInstancePerFile {
-    fn on_query_match(&self, listener_index: usize, node: Node, context: &mut QueryMatchContext) {
-        match listener_index {
-            0 => {
+fn create_identifier_replacing_rule(
+    name: impl Into<String>,
+    replacement: impl Into<String>,
+) -> Arc<dyn Rule> {
+    rule! {
+        name => format!("replace_{}_with_{}", self.name, self.replacement),
```

Comment:
Ok so the `self.name`/`self.replacement` here wouldn't be getting re-written, they will be referencing the actual "rule-static" `name`/`replacement` fields (as specified below in the `state` section) on the `Rule`-implementing struct

(and this expression gets run in a context where that `self` is visible, specifically inside of `impl Rule for ... { ... }`

And the `name`/`replacement` fields "have to" be "rule-static" because they get initialized to values that are "closed-over" from the place where `rule!()` was called (and everything has to be `'static` (or at least that's what it seemed like made sense))

Actually since the usage pattern is that these rule types are getting defined "inline in an expression context", maybe it would be just as easy/valid for non-"rule-static" state fields to get initialized from "closed-over" values too?

I could try that and look at the generated code

A little hard to wrap my head around "what's happening where/when"

But the goal would be hopefully that there would be a coherent mental model such that authoring rules could feel pretty fluid